add: separate class for sdk scripts

itsspriyansh opened the pull request.

garg3133 commented:
Initial thought: Let's have a new `commands/` sub-directory inside the `android/` directory and arrange all our sub-commands related to `android` inside that.

For example, keep the code related to the `connect` sub-command inside `android/commands/connect/` directory, etc.

While for the `dot` commands (ex. `android.adb`, `android.emulator`, etc.) we can put them inside a `dotcommands.ts` file directly inside `android/` since the code for the dot commands should be relatively straightforward as we just need to forward the arguments to the actual binaries here.

itsspriyansh commented:
@garg3133 This approach sounds pretty good. I have moved the script according to this new folder structure. Force pushed due to rebase.

garg3133 commented on src/index.ts:
```diff
@@ -18,12 +19,17 @@ export const run = () => {
-      androidSetup.run();
-    } else {
+      if (args[1]) {
+        const sdkCommandExecute = new SdkCommandExecute(args[1], options);
```
Maybe rename the class to `AndroidSubcommand`? And so the next line would be `androidSubcommand.run()`.

garg3133 commented on src/commands/android/commands/connect/connect-wireless-adb.ts:
Let's rename this to `commands/android/subcommands/connect/wireless.ts` instead.

Having `commands` twice seems confusing in the file tree.

garg3133 commented on src/commands/android/execute-commands.ts:
Rename this to `commands/android/subcommands/index.ts`

itsspriyansh replied:
Just wanted to confirm, did you mean `commands/android/commands/index.ts`?

garg3133 commented on src/commands/android/execute-commands.ts:
```diff
@@ -0,0 +1,82 @@
+
+  async run(): Promise<boolean> {
+    if (!Object.keys(AVAILABLE_SUBCOMMANDS).includes(this.subcommand)) {
+      showHelp([], this.subcommand);
```
Maybe do a separate implementation of `showHelp` for subcommands? The help content for the subcommands is anyways an extension to the original help content and can be pulled out as a separate function. We don't need to force-implement everything into a single function if splitting the implementation into different functions makes more sense.

The main motive here is to show help at every step to the users that makes sense to them. So, the help section at every different level of the project would be different.

For ex., if you run `npx @nightwatch/mobile-helper andr`, it will show you a very minimal help section with all the commands supported. And if you go one level inside and make some mistake in entering the options for the `android` command (instead of the `android` command itself), it will show you help related to the options. 

But since we are integrating the subcommands with the `android` command itself, the latter case should also show all the available subcommands along with all the available options (in case user actually wanted to run a subcommand), as you correctly did in the other PR. But instead of merging everything together in a single function, we can split them out into two (or more) functions, each handling a specific section of the help section. And the division of these sections would be in terms of usage:

* Usage 1: Verify if all the requirements are met to run tests on an Android device/emulator.
  > Display all available options to verify/setup.
* Usage 2: Control and manage your already existing Android SDK setup using subcommands.
  > Display all available subcommands with major options (not all of them).

Now, once this is done, we might need to require a separate Help for each sub-commands, showing the options for that one subcommand only. This is for the case when the user got the subcommand right but messed up with the options. The implementation could be done in the `android/subcommands/<command-name>/index.ts` file. 

But remember not to duplicate any data for these individual cases. For ex., the data for all the available subcommands and their corresponding options should come from a single place or constant, and we'll use that data differently in different help section implementations.

garg3133 replied:
Please let me know if there's anything that don't make sense here.

itsspriyansh replied:
@garg3133 Thanks. I understood it pretty well. As for dividing the `showHelp` in seperate functions, will it be right to keep the current `showHelp` in `AndroidSetup` as it is and create separate function for subcommands help and invoke it inside both current `showHelp` function and inside the `AndroidSubcommand` class? Its because the current `showHelp` is not required anywhere else and will be always called in just one place, that is inside `AndroidSetup`.

For eg. 
Lets create `showSubcommandHelp` and put it inside `common.ts`.
1. If user passes `npx @nightwatch/mobile-helper android -h` then `showHelp` inside `AndroidSetup` will execute and display all options. Then it will invoke `showSubcommandHelp` and display help for available subcommands.
2. If user passes `npx @nightwatch/mobile-helper android random` then only `showSubcommandHelp` will be called inside `AndroidSubcommand` class.

garg3133 replied:
Makes sense.

garg3133 commented on src/commands/android/execute-commands.ts:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  async executeSdkScript(): Promise<boolean> {
+    if (this.subcommand === "connect") {
+    // check the corresponding options for the 'connect' subcommand.
+      if (this.options.wireless) {
+        // execute script for wireless adb connection.
+        return await connectWirelessAdb(this.sdkRoot, this.platform);
+      }
```
This file should only determine if the subcommand is `connect` and if so, move the flow to `subcommands/connect/index.ts` file that would handle the rest of the logic related to the `connect` subcommand.

itsspriyansh commented:
@garg3133 I have changed the filenames and fixed ESLint errors. 
I have also removed anything related to `showHelp` from this PR. That will be handled in the next PR.
For now, this PR will just move the wireless connection script and add the new subcommand class with basic functionalities. Also minor changes in `src/index.ts`.

garg3133 commented on src/index.ts:
```diff
@@ -4,6 +4,7 @@ import minimist from 'minimist';
 import {AndroidSetup} from './commands/android';
 import {IosSetup} from './commands/ios';
 import {AVAILABLE_COMMANDS} from './constants';
+import {AndroidSubcommand} from './commands/android/subcommands/index';
```
No need to mention `/index`.

```suggestion
import {AndroidSubcommand} from './commands/android/subcommands';
```

garg3133 commented on src/index.ts:
```diff
@@ -18,12 +19,17 @@ export const run = () => {
+        const androidSubcommand = new AndroidSubcommand(args[1], options);
+        androidSubcommand.run();
+      } else {
+        const androidSetup = new AndroidSetup(options, args[1]);
```
Why pass `args[1]` here?

garg3133 commented on src/index.ts:
```diff
@@ -33,14 +39,16 @@ export const run = () => {
-  if (cmdPassed) {
-    console.log(colors.red(`unknown command: ${cmdPassed}`), '\n');
-  } else if (!helpFlag) {
+const showHelp = (args: string[], helpFlag: boolean) => {
+  if (args.length > 2) {
+    console.log(colors.red('Too many arguments passed.'), '\n');
+  } else if (args[0]) {
+    console.log(colors.red(`unknown command: ${args[0]}`), '\n');
+  } else if (!args[0] && !helpFlag) {
```
The logic here is getting confusing and requires the context of the code calling this function to fully understand what's going on in this. By just looking at this function you would not understand why it is printing an error if `arg[0]` is present, for example, which is not ideal.

Understanding a function's logic should not require us to know the context of the code from where it is called. So, we should probably move these error messages to the `if-else` block itself in the `run` function. Also, split the `|| args.length > 2` clause into a separate `else if` block for better readability.

garg3133 replied:
Also, instead of just checking `args.length > 2`, we should probably also check `args[0] === 'ios' && args.length > 1`, so the new `else if` statement would be:

```ts
} else if (args.length > 2 || (args[0] === 'ios' && args.length > 1)) {
```

garg3133 commented on src/index.ts:
```diff
@@ -18,12 +19,17 @@ export const run = () => {
-      const androidSetup = new AndroidSetup(options);
-      androidSetup.run();
-    } else {
+      if (args[1]) {
+        const androidSubcommand = new AndroidSubcommand(args[1], options);
```
```suggestion
      if (args[1]) {
        // args[1] represents an android subcommand
        const androidSubcommand = new AndroidSubcommand(args[1], options);
```

garg3133 commented on src/commands/android/subcommands/index.ts:
```diff
@@ -0,0 +1,65 @@
+    if (!sdkRootEnv) {
+      Logger.log(colors.red('Path to Android SDK not found in environment variables!\n'));
+      Logger.log(`Run: ${colors.cyan('npx @nightwatch/mobile-helper android --standalone')} to setup Android SDK`);
+      Logger.log('     or provide the path to Android SDK if already installed.');
```
why this line? where would they provide the path to Android SDK? I think running the setup command would already handle this and they would need to provide the Android SDK path there.

garg3133 commented on src/commands/android/subcommands/index.ts:
```diff
@@ -0,0 +1,65 @@
+    }
+    this.sdkRoot = sdkRootEnv;
+
+    this.executeSdkScript();
```
```suggestion
    this.executeSubcommand();
```

garg3133 commented:
The overall architecture and folder structure for subcommands looks good now! Good job!!

itsspriyansh commented:
@garg3133 I have resolved everything as you suggested. I have removed the error message when `ANDROID_HOME` is not found in environment variables because the function `getSdkRootFromEnv` already does that. The comment below that was meant to inform user that they can provide the path using the same command as above, however I don't think that is necessary so I have removed that as well.

garg3133 commented on src/commands/android/subcommands/index.ts:
```diff
@@ -0,0 +1,63 @@
+      return false;
+    }
+
+    const sdkRootEnv = getSdkRootFromEnv(this.androidHomeInGlobalEnv, this.rootDir);
```
Swap the arguments here as well.

garg3133 commented:
> The comment below that was meant to inform user that they can provide the path using the same command as above, however I don't think that is necessary so I have removed that as well.

Now I get your concern and it makes sense because when we say "Run: `npx @nightwatch/mobile-helper android --standalone` to setup Android SDK", it might confuse users if they've already setup the Android SDK requirements.

In that case, I think we can just give them the command to resolve the issue instead giving them any extra information on what they can do or can't do using that command. For example, we can just keep it the instruction short and simple like below:

<img width="530" alt="image" src="https://github.com/nightwatchjs/mobile-helper-tool/assets/39924567/c60c1ac9-037a-4ab3-bba0-5251c812b137">

garg3133 commented:
The help section is still not fully refined. For instance, if we run the following:

<img width="440" alt="image" src="https://github.com/nightwatchjs/mobile-helper-tool/assets/39924567/fc58ade7-36d3-4055-96b5-5a9dcd5d77e0">

Even though we passed the "android" command correctly, the help is very generic whereas we should ask them to run `npx @nightwatch/mobile-helper android --help` to know more about how to use the android command.

So, the "Too many arguments passed" error does not make much sense here. Instead, if the main command is correct, we should probably just give it to the android section and let it handle.

And on another thought, we should probably just create an `index.ts` file inside `commands/android` as well (just like we did with `commands/android/subcommands`) to handle all this stuff, including the logic for upcoming dot commands.
(Move the existing `android/index.ts` file to `android/androidSetup.ts` and then create a new `index.ts` file as stated above.)

But none of this needs to be done in the current PR, this looks good and can be merged as soon as you handle the above two comments.

itsspriyansh commented:
@garg3133 Thanks for pointing out the issue with `showHelp` function. I will be looking into it. 
For now I have updated this PR with the other two changes you suggested along with some other minor refactors.

garg3133 commented:
Looks good now, thanks!
